### PR TITLE
Keep guest SF/ZF/CF/OF flags resident in host NZCV

### PR DESCRIPTION
**Pull request by alyssarosenzweig**
This adds all the infrastructure needed to take advantage of flag manipulation instructions.

### CODE REVIEW

**Comment by Sonicadvance1** [comment]
So I see your problem but I don't have a solution yet.
Retaining flags in ExitFunction is nice but we aren't guaranteed to jump directly to another JIT block or potentially back to the assembly dispatcher. https://github.com/FEX-Emu/FEX/blob/main/FEXCore/Source/Interface/Core/Dispatcher/Arm64Dispatcher.cpp#L86
If it makes it to the assembly dispatcher then it is going to stomp NZCV state.

Saving to CoreState in ExitFunction and restoring when entering a codeblock didn't entirely solve the problem though, so something else must still be clobbering it.

**Comment by Sonicadvance1** [comment]
Additionally the CoreState flags initialization for the NZCV/AF/PF registers might not be initialized correctly for default?
I think this was technically a problem once a flag was being stored inverted from x86 representation?
Didn't look too closely at that, by default all the flags should initialize to zero representation. Unlikely to be related

**Comment by Sonicadvance1** [comment]
Convenience bugfix in a comment!
```diff
ryanh@ubuntu-linux-20-04-desktop:~/work/FEXNew/FEXCore$ git diff -w
diff --git a/FEXCore/Source/Interface/Core/Dispatcher/Arm64Dispatcher.cpp b/FEXCore/Source/Interface/Core/Dispatcher/Arm64Dispatcher.cpp
index 7413309cb..4e2224cec 100644
--- a/FEXCore/Source/Interface/Core/Dispatcher/Arm64Dispatcher.cpp
+++ b/FEXCore/Source/Interface/Core/Dispatcher/Arm64Dispatcher.cpp
@@ -98,8 +98,9 @@ void Arm64Dispatcher::EmitDispatcher() {
     and_(ARMEmitter::Size::i64Bit, ARMEmitter::Reg::r3, RipReg.R(), LookupCache::L1_ENTRIES_MASK);
     add(ARMEmitter::Size::i64Bit, ARMEmitter::Reg::r0, ARMEmitter::Reg::r0, ARMEmitter::Reg::r3, ARMEmitter::ShiftType::LSL , 4);
     ldp<ARMEmitter::IndexType::OFFSET>(ARMEmitter::XReg::x3, ARMEmitter::XReg::x0, ARMEmitter::Reg::r0, 0);
-  cmp(ARMEmitter::Size::i64Bit, ARMEmitter::Reg::r0, RipReg.R());
-  b(ARMEmitter::Condition::CC_NE, &FullLookup);
+
+    sub(ARMEmitter::XReg::x0, ARMEmitter::XReg::x0, RipReg);
+    cbnz(ARMEmitter::Size::i64Bit, ARMEmitter::Reg::r0, &FullLookup);

     br(ARMEmitter::Reg::r3);

@@ -144,8 +145,8 @@ void Arm64Dispatcher::EmitDispatcher() {
       ldp<ARMEmitter::IndexType::OFFSET>(ARMEmitter::XReg::x3, ARMEmitter::XReg::x1, ARMEmitter::Reg::r0, 0);

       // If the guest address doesn't match, Compile the block.
-    cmp(ARMEmitter::XReg::x1, RipReg);
-    b(ARMEmitter::Condition::CC_NE, &NoBlock);
+      sub(ARMEmitter::XReg::x1, ARMEmitter::XReg::x1, RipReg);
+      cbnz(ARMEmitter::Size::i64Bit, ARMEmitter::Reg::r1, &NoBlock);

       // Check the host address to see if it matches, else compile the block.
       cbz(ARMEmitter::Size::i64Bit, ARMEmitter::Reg::r3, &NoBlock);
@@ -201,7 +202,7 @@ void Arm64Dispatcher::EmitDispatcher() {
       FillStaticRegs();

     ldr(ARMEmitter::XReg::x1, STATE, offsetof(FEXCore::Core::CPUState, DeferredSignalRefCount));
-    subs(ARMEmitter::Size::i64Bit, ARMEmitter::XReg::x1, ARMEmitter::XReg::x1, 1);
+    sub(ARMEmitter::Size::i64Bit, ARMEmitter::XReg::x1, ARMEmitter::XReg::x1, 1);
     str(ARMEmitter::XReg::x1, STATE, offsetof(FEXCore::Core::CPUState, DeferredSignalRefCount));

     // Trigger segfault if any deferred signals are pending
@@ -237,7 +238,7 @@ void Arm64Dispatcher::EmitDispatcher() {
       FillStaticRegs();

     ldr(ARMEmitter::XReg::x0, STATE, offsetof(FEXCore::Core::CPUState, DeferredSignalRefCount));
-    subs(ARMEmitter::Size::i64Bit, ARMEmitter::XReg::x0, ARMEmitter::XReg::x0, 1);
+    sub(ARMEmitter::Size::i64Bit, ARMEmitter::XReg::x0, ARMEmitter::XReg::x0, 1);
     str(ARMEmitter::XReg::x0, STATE, offsetof(FEXCore::Core::CPUState, DeferredSignalRefCount));

     // Trigger segfault if any deferred signals are pending
```

File: Source/Tools/FEXLoader/HarnessHelpers.h [hunk 1]
```diff
@@ -405,11 +405,7 @@ namespace FEX::HarnessHelper {
         return Result;
       };
 
-#ifndef _WIN32
-      const auto AllocPageSize = FHU::FEX_PAGE_SIZE;
```
Comment (Sonicadvance1) [inline Source/Tools/FEXLoader/HarnessHelpers.h:409]: This would probably be fine to just change to using `getpagesize()` since it is low impact here btw.

Comment (alyssarosenzweig) [reply Source/Tools/FEXLoader/HarnessHelpers.h:409]: ack. this is just for me but sure :clown_face:

File: FEXCore/Source/Interface/IR/IR.json [hunk 2]
```diff
@@ -1899,10 +1920,12 @@
                  "Each element is compared, if the result is true then the resulting element is ~0, else zero"
                 ],
 
+        "ImplicitFlagClobber": true,
         "DestSize": "RegisterSize",
         "NumElements": "RegisterSize / ElementSize"
       },
       "FPR = VFCMPEQ u8:#RegisterSize, u8:#ElementSize, FPR:$Vector1, FPR:$Vector2": {
+        "ImplicitFlagClobber": true,
```
Comment (Sonicadvance1) [inline FEXCore/Source/Interface/IR/IR.json:1928]: This one is an oversight. VFCMPEQ uses fcmeq for SVE which doesn't set flags. Which makes the MMX pfcmpeq worse.

Comment (alyssarosenzweig) [reply FEXCore/Source/Interface/IR/IR.json]: nice, fixed!

File: FEXCore/Source/Interface/Core/ArchHelpers/CodeEmitter/ALUOps.inl [hunk 3]
```diff
@@ -766,6 +766,11 @@ public:
     EvaluateIntoFlags(Op, 1, rn);
   }
 
+  void cfinv() {
```
Comment (Sonicadvance1) [inline FEXCore/Source/Interface/Core/ArchHelpers/CodeEmitter/ALUOps.inl:769]: This needs a unittest in `FEXCore/unittests/Emitter/ALU_Tests.cpp` probably near the setf16 test.

**Comment by Sonicadvance1** [comment]
Tried giving this a test, looks like there are still some bugs somewhere.
Disabling optimization passes was necessary for steam to get running.
Needed to change the block size down to 1 instruction for Ender Lilies not to crash.

So /something/ is unhappy.

**Comment by Sonicadvance1** [comment]
PR #3222 opened to showcase a bug that this PR has. My current assumption is that any instruction that generates multiple blocks of code inside of OpcodeDispatcher has a chance of breaking NZCV flags.

**Comment by Sonicadvance1** [comment]
Poked some more at this. Running winecfg also seemingly doesn't work. I have isolated this one down to the inline syscall handling for futexes. Skipping inlining the futex syscall gets winecfg working at least, this is implying NZCV corruption around futex. I guess this is from interrupts with the futex? I'll generate some unittests around this and see if I can capture it more clearly.
It's sadly not enough to get Ender Lilies running, so there must be more than just this issue. Maybe a chance for nzcv corruption from any signal? We save and restore pstate on signal but if that's the case then we're missing something.

**Comment by Sonicadvance1** [comment]
Unfortunately my local unittest that I wrote doesn't uncover any issue in this PR. It did find some instability in 32-bit signal handling in main, but that's unrelated to this.

Will need to poke at this harder.

**Comment by alyssarosenzweig** [comment]
Thanks to @Sonicadvance1's debugging, this should be working now :tada: 

I expect a perf **regression** from this PR in its current state due to needing to rework branches and selects. That will follow soon. Since we just cut the 2311 release and I expect to complete that work by 2312, I think it makes sense to merge in this form, provided this is functionally correct.

PR on top of #3257 (already reviewed and written for this), either that can land first or we can just land it altogether here.

Le's go!

**Comment by Sonicadvance1** [comment]
I'll poke this a bit later tonight to see if there are any remaining blockers.

**Comment by Sonicadvance1** [comment]
Looks like there is still some more work to go, Ender Lilies still explodes in this PR. Time to bring back the check branch

**Comment by Sonicadvance1** [comment]
Threw about 100 games at this and didn't notice a regression. I'll do a code review on this later today.

File: unittests/InstructionCountCI/FlagM/Secondary.json [hunk 4]
```diff
@@ -2250,13 +2321,11 @@
...
-        "cmn x7, x20",
-        "mrs x20, nzcv",
-        "str w20, [x28, #728]"
+        "cmn x7, x20"
       ]
     },
     "pmovmskb eax, mm0": {
-      "ExpectedInstructionCount": 12,
+      "ExpectedInstructionCount": 14,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/FlagM/Secondary.json:2329]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/FlagM/Secondary.json [hunk 5]
```diff
@@ -2266,25 +2335,29 @@
...
         "addp v2.8b, v2.8b, v2.8b",
-        "umov w4, v2.h[0]"
+        "umov w4, v2.h[0]",
+        "msr nzcv, x20"
       ]
     },
     "maskmovq mm0, mm1": {
-      "ExpectedInstructionCount": 6,
+      "ExpectedInstructionCount": 8,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/FlagM/Secondary.json:2350]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 6]
```diff
@@ -999,12 +1052,14 @@
       ]
     },
     "minps xmm0, xmm1": {
-      "ExpectedInstructionCount": 2,
+      "ExpectedInstructionCount": 4,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:1056]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 7]
```diff
@@ -1016,12 +1071,14 @@
       ]
     },
     "maxps xmm0, xmm1": {
-      "ExpectedInstructionCount": 2,
+      "ExpectedInstructionCount": 4,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:1075]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 8]
```diff
@@ -1126,36 +1183,42 @@
       ]
     },
     "pcmpgtb mm0, mm1": {
-      "ExpectedInstructionCount": 4,
+      "ExpectedInstructionCount": 6,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:1187]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 9]
```diff
@@ -1126,36 +1183,42 @@
...
         "cmgt v2.16b, v3.16b, v2.16b",
-        "str d2, [x28, #752]"
+        "str d2, [x28, #752]",
+        "msr nzcv, x20"
       ]
     },
     "pcmpgtw mm0, mm1": {
-      "ExpectedInstructionCount": 4,
+      "ExpectedInstructionCount": 6,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:1200]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 10]
```diff
@@ -1126,36 +1183,42 @@
...
         "cmgt v2.8h, v3.8h, v2.8h",
-        "str d2, [x28, #752]"
+        "str d2, [x28, #752]",
+        "msr nzcv, x20"
       ]
     },
     "pcmpgtd mm0, mm1": {
-      "ExpectedInstructionCount": 4,
+      "ExpectedInstructionCount": 6,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:1213]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 11]
```diff
@@ -1346,36 +1409,42 @@
       ]
     },
     "pcmpeqb mm0, mm1": {
-      "ExpectedInstructionCount": 4,
+      "ExpectedInstructionCount": 6,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:1413]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 12]
```diff
@@ -1346,36 +1409,42 @@
...
         "cmeq v2.16b, v3.16b, v2.16b",
-        "str d2, [x28, #752]"
+        "str d2, [x28, #752]",
+        "msr nzcv, x20"
       ]
     },
     "pcmpeqw mm0, mm1": {
-      "ExpectedInstructionCount": 4,
+      "ExpectedInstructionCount": 6,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:1426]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 13]
```diff
@@ -1346,36 +1409,42 @@
...
         "cmeq v2.8h, v3.8h, v2.8h",
-        "str d2, [x28, #752]"
+        "str d2, [x28, #752]",
+        "msr nzcv, x20"
       ]
     },
     "pcmpeqd mm0, mm1": {
-      "ExpectedInstructionCount": 4,
+      "ExpectedInstructionCount": 6,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:1439]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 14]
```diff
@@ -4169,7 +4254,7 @@
       ]
     },
     "pmovmskb eax, mm0": {
-      "ExpectedInstructionCount": 12,
+      "ExpectedInstructionCount": 14,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:4258]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/Secondary.json [hunk 15]
```diff
@@ -4526,16 +4613,18 @@
       ]
     },
     "maskmovq mm0, mm1": {
-      "ExpectedInstructionCount": 6,
+      "ExpectedInstructionCount": 8,
       "Optimal": "Yes",
       "Comment": "0x0f 0xf7",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/Secondary.json:4618]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/VEX_map1.json [hunk 16]
```diff
@@ -1030,72 +1030,84 @@
       ]
     },
     "vpcmpgtb xmm0, xmm1, xmm2": {
-      "ExpectedInstructionCount": 1,
+      "ExpectedInstructionCount": 3,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/VEX_map1.json:1034]: Needs to be changed to `No`. And the rest of the vpcmp effected in this file.

File: unittests/InstructionCountCI/VEX_map1.json [hunk 17]
```diff
@@ -5785,16 +5832,18 @@
       ]
     },
     "vmaskmovdqu xmm0, xmm1": {
-      "ExpectedInstructionCount": 4,
+      "ExpectedInstructionCount": 6,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/VEX_map1.json:5836]: Needs to be changed to `No`.

File: unittests/InstructionCountCI/VEX_map2.json [hunk 18]
```diff
@@ -1032,26 +1044,30 @@
       ]
     },
     "vpcmpeqq xmm0, xmm1, xmm2": {
-      "ExpectedInstructionCount": 1,
+      "ExpectedInstructionCount": 3,
       "Optimal": "Yes",
```
Comment (Sonicadvance1) [inline unittests/InstructionCountCI/VEX_map2.json:1048]: Needs to be changed to `No`. And the rest of the vpcmp effected in this file.

**Review by Sonicadvance1 — COMMENTED** [review]
Initial review pass.
- Can rebase and drop the ender lilies commit now.
- Interesting some code-gen like adc and sbb and the lock variants will be able to be easily improved after this
- Some cute redundant code emitted in some instructions like sahf
- Some cute code improvements, like cmc becoming optimal
- Some not so cute, like vector mask generation instructions becoming non-optimal for seemingly no reasons? Marked all these with comments
- cmov/fcmov/cset improvements in follow-up PRs will be nice.
- Some of the repeat instructions spill nzcv and then fill it for seemingly no reason, which is interesting.

Overall very exciting! I don't think all perf regressions need to be fixed before this PR is merged, but maybe some of the easier ones would be nice, like the vector mask operations.

I'll do some benchmarking of this to make sure we don't regress performance too much here, but so far I don't see much of an issue. Especially with the instruction count reduction on most ALU operations.

**Comment by alyssarosenzweig** [comment]
>     * Can rebase and drop the ender lilies commit now.
Done.

>     * Some not so cute, like vector mask generation instructions becoming non-optimal for seemingly no reasons? Marked all these with comments
Did the ugly trick to avoid implicit clobber on all of the relevant operations, so none of these regress code gen now :smile: 
 
>     * cmov/fcmov/cset improvements in follow-up PRs will be nice.
Yep, that's up next.
 
>     * Some of the repeat instructions spill nzcv and then fill it for seemingly no reason, which is interesting.
:woman_shrugging:

**Comment by alyssarosenzweig** [comment]
the rep cmps silliness will go away when I redo CondJump, which is probably 2nd in line (1st is Select)

**Review by Sonicadvance1 — APPROVED** [review]
:tada:

**Pull request merged**